File: go/bob/bob.go
// Package bob should have a package comment that summarizes what it's about.
package bob

import (
	"regexp"
	"strings"
)

// Hey should have a comment documenting it.
func Hey(remark string) string {
	if isNoLetters(remark) {
		return "Whatever."
	}
	if isForcefulQuestion(remark) {
		return "Calm down, I know what I'm doing!"
	}
	if isLoud(remark) {
		return "Whoa, chill out!"
	}
	if isQuestion(remark) {
		return "Sure."
	}
	return "Whatever."
}

func isNoLetters(remark string) bool {
	// return regexp.MustCompile(`^[0-9]+$`).MatchString(remark)
	return regexp.MustCompile(`(\d+)(,\s*\d+)*`).MatchString(remark)
}

func isForcefulQuestion(remark string) bool {
	return isLoud(remark) && isQuestion(remark)
}

func isExcited(remark string) bool {
	return strings.HasSuffix(remark, "!")
}

func isLoud(remark string) bool {
	return strings.ToUpper(remark) == remark
}

func isQuestion(remark string) bool {
	return strings.HasSuffix(remark, "?")
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      empty?(input) ->
        "Fine. Be that way!"

      shouting_questions?(input) ->
        "Calm down, I know what I'm doing!"

      question?(input) ->
        "Sure."

      shouting?(input) ->
        "Whoa, chill out!"

      true ->
        "Whatever."
    end
  end

  defp empty?(input) do
    String.trim(input) == ""
  end

  defp question?(input) do
    ends_with?("?", input)
  end

  defp shouting?(input) do
    String.upcase(input) == input && String.downcase(input) != input
  end

  def shouting_questions?(input) do
    shouting?(input) && question?(input)
  end

  defp ends_with?(char, input) do
    String.last(input) == char
  end
end
